plugins/k8saudit/pkg/k8saudit/extract.go

```diff
@@ -0,0 +1,733 @@
+	ErrExtractWrongType = fmt.Errorf("wrong type conversion")
+	//
+	// ErrExtractBrokenJSON indicates that the requested field cannot be
+	// extracted from a certain event due to the internal JSON prepresentaiton
+	// being broken or badly formatted
+	ErrExtractBrokenJSON = fmt.Errorf("broken JSON data")
+	//
+	// ErrExtractUnsupportedType indicates that the requested field cannot be
+	// extracted from a certain event due to its field type being not supported
+	ErrExtractUnsupportedType = fmt.Errorf("type not supported")
+)
+
+// AuditEventExtractor is an helper that extract K8S Audit fields from
+// K8S Audit events. The event data is expected to be a JSON that in the form
+// that is provided by K8S Audit webhook (see https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#webhook-backend).
+// The ExtractFromEvent method can be used to easily process an ExtractRequest.
+// If the Audit Event data is nested inside another JSON object, you can use
+// a combination of the Decode/DecodeEvent and ExtractFromJSON convenience
+// methods. AuditEventExtractor relies on the fastjson package for performanct
+// manipulation of JSON data.
+type AuditEventExtractor struct {
+	jparser     fastjson.Parser
+	jdata       *fastjson.Value
+	jdataEvtnum uint64
+}
+
+// Decode parses a JSON value from a io.Reader
+func (e *AuditEventExtractor) Decode(evtNum uint64, reader io.ReadSeeker) (*fastjson.Value, error) {
+	// as a very quick sanity check, only try to extract all if
+	// the first character is '{' or '['
+	data := []byte{0}
+	_, err := reader.Read(data)
+	if err != nil {
+		return nil, err
+	}
+	if !(data[0] == '{' || data[0] == '[') {
+		return nil, ErrExtractBrokenJSON
+	}
+	// decode the json, but only if we haven't done it yet for this event
+	if evtNum != e.jdataEvtnum {
+		_, err := reader.Seek(0, io.SeekStart)
+		if err != nil {
+			return nil, err
+		}
+		data, err = ioutil.ReadAll(reader)
+		if err != nil {
+			return nil, err
+		}
+		e.jdata, err = e.jparser.ParseBytes(data)
+		if err != nil {
+			return nil, err
+		}
+		e.jdataEvtnum = evtNum
+	}
+	return e.jdata, nil
+}
+
+// DecodeEvent parses a JSON value from a sdk.EventReader
+func (e *AuditEventExtractor) DecodeEvent(evt sdk.EventReader) (*fastjson.Value, error) {
+	return e.Decode(evt.EventNum(), evt.Reader())
+}
+
+// ExtractFromEvent processes a sdk.ExtractRequest and extracts a
+// field by reading data from a sdk.EventReader
+func (e *AuditEventExtractor) ExtractFromEvent(req sdk.ExtractRequest, evt sdk.EventReader) error {
+	jsonValue, err := e.DecodeEvent(evt)
+	if err != nil {
+		return err
+	}
+	return e.ExtractFromJSON(req, jsonValue)
+}
+
+// ExtractFromJSON processes a sdk.ExtractRequest and extracts a
+// field by reading data from a jsonValue *fastjson.Value
+func (e *AuditEventExtractor) ExtractFromJSON(req sdk.ExtractRequest, jsonValue *fastjson.Value) error {
+	switch req.Field() {
+	case "ka.auditid":
+		return e.extractFromKeys(req, jsonValue, "auditID")
```

Comment:
That's a very great point. My feeling is that JSON Pointers would not help much here, mostly because they would add unnecessary overhead in parsing the pointer to an explorable struct. Plus, I haven't personally found any popular JSON pointer package for golang.

As for that performance comparison, it looks like `jsonparser` might be the way to go for our use case. `go-jmespath` probably offers better query capabilities, but it seems like it's not strong on performance. Overall, I feel like it would be good for us to create an internal interface for JSON-related access to share across all plugins, so that it will be easier to upgrade/swap the underlying package in the future. However, it's hard to predict all the possible use cases we'll have in the future and understand how to shape such interface.

Reply:
IMHO, this PR is big enough. So we should implement any other optimization afterward.